### PR TITLE
Support Apache HttpClient 5.2+

### PR DESCRIPTION
**Pull request by kojilin**
* Support classic/async/h2async client
* Compare to 4.x, 5.x exposes ExecInterceptor. So we can avoid making a builder class and just add ExecInterceptor/RequestInterceptor/ResponseInterceptor to achieve the tracing.

--
### Why 5.2?

* 5.1.x has issues which makes our IT for async failed by open/close in different thread.(Not investigate yet but 5.0.4 & 5.2 doesn't have this issue)
```
java.io.IOException: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Thread [Time-limited test] opened scope, but thread [httpclient-dispatch-1] closed it
	at brave.httpclient5.ITTracingH2AsyncClientBuilder.blockOnFuture(ITTracingH2AsyncClientBuilder.java:54)
	at brave.httpclient5.ITTracingH2AsyncClientBuilder.invoke(ITTracingH2AsyncClientBuilder.java:47)
	at brave.httpclient5.ITTracingH2AsyncClientBuilder.get(ITTracingH2AsyncClientBuilder.java:87)
	at brave.httpclient5.ITTracingH2AsyncClientBuilder.get(ITTracingH2AsyncClientBuilder.java:44)
	at brave.test.http.ITHttpClient.httpPathTagExcludesQueryParams(ITHttpClient.java:387)
```

* 5.0.4 has different that ITTracingH2AsyncClientBuilder#failedRequestInterceptorRemovesScope throws RuntimeException directly instead of `java.io.IOException: java.util.concurrent.ExecutionException: java.lang.RuntimeException: Test`. 5.0.3, 5.2 doesn't have this issue.

* 5.0.3 has issues which make H2Async & CachingH2Async not work(test will be failed)
  * https://issues.apache.org/jira/browse/HTTPCLIENT-2123
  * https://issues.apache.org/jira/browse/HTTPCLIENT-2126 

--
Solve https://github.com/openzipkin/brave/issues/1228

### CODE REVIEW

File: instrumentation/httpclient5/src/main/java/brave/httpclient5/AsyncHandleSendHandler.java
Comment (adriancole) [inline instrumentation/httpclient5/src/main/java/brave/httpclient5/AsyncHandleSendHandler.java:27]: this seems we should fix httpTracing.serverName() to not permit empty string.. mind moving it there? (ex builder coerce empty to null)

File: instrumentation/httpclient5/pom.xml
Comment (adriancole) [inline instrumentation/httpclient5/pom.xml:32]:
```suggestion
    <module.name>brave.httpclient5</module.name>
```

File: instrumentation/httpclient5/pom.xml
Comment (adriancole) [inline instrumentation/httpclient5/pom.xml:35]: double check if this is the minimum version of httpclient5?

File: instrumentation/httpclient5/src/main/java/brave/httpclient5/HandleReceiveHandler.java
Comment (adriancole) [inline instrumentation/httpclient5/src/main/java/brave/httpclient5/HandleReceiveHandler.java:36]:
```suggestion
class HandleReceiveHandler implements ExecChainHandler {
```

File: instrumentation/httpclient5/src/main/java/brave/httpclient5/HandleReceiveHandler.java
Comment (adriancole) [inline instrumentation/httpclient5/src/main/java/brave/httpclient5/HandleReceiveHandler.java:37]: make all these package private as bytecode is smaller and it is more legible

File: instrumentation/httpclient5/src/main/java/brave/httpclient5/HandleSendHandler.java
Comment (adriancole) [inline instrumentation/httpclient5/src/main/java/brave/httpclient5/HandleSendHandler.java:33]:
```suggestion
class HandleSendHandler implements ExecChainHandler {
```

File: instrumentation/httpclient5/src/main/java/brave/httpclient5/HandleSendHandler.java
Comment (adriancole) [inline instrumentation/httpclient5/src/main/java/brave/httpclient5/HandleSendHandler.java:39]:
```suggestion
  HandleSendHandler(HttpTracing httpTracing) {
```

File: instrumentation/httpclient5/src/main/java/brave/httpclient5/HttpClientUtils.java
Comment (adriancole) [inline instrumentation/httpclient5/src/main/java/brave/httpclient5/HttpClientUtils.java:11]:
```suggestion
class HttpClientUtils {
```

File: instrumentation/httpclient5/src/main/java/brave/httpclient5/TracingHttpClients.java
Comment (adriancole) [inline instrumentation/httpclient5/src/main/java/brave/httpclient5/TracingHttpClients.java:30]:
```suggestion
public class HttpClient5Tracing {
```

File: instrumentation/httpclient5/src/main/java/brave/httpclient5/TracingHttpClients.java
Comment (adriancole) [inline instrumentation/httpclient5/src/main/java/brave/httpclient5/TracingHttpClients.java:31]: separate duty of creating HttpClient5Tracing component from making a CloseableHttpClient. this allows one-time configuration even if using multiple components.

ex.
```java
HttpClient5Tracing httpClient5Tracing = HttpClient5Tracing.newBuilder(httpTracing);

httpAsyncClient = httpClient5Tracing.build(httpAsyncClientBuilder);
```

File: instrumentation/httpclient5/src/test/java/brave/httpclient5/ITTracingCachingH2ClientBuilder.java
Comment (adriancole) [inline instrumentation/httpclient5/src/test/java/brave/httpclient5/ITTracingCachingH2ClientBuilder.java:30]:
```suggestion
class ITTracingCachingH2ClientBuilder extends ITTracingH2AsyncClientBuilder {
```

File: instrumentation/httpclient5/src/test/java/brave/httpclient5/ITTracingH2AsyncClientBuilder.java
Comment (adriancole) [inline instrumentation/httpclient5/src/test/java/brave/httpclient5/ITTracingH2AsyncClientBuilder.java:42]: just be lazy and throw new IOException(e) for wide catch Exception as it is too much distraction in the test file and outcome would be similar as long as surefire is set to include stacktrace (which it is)

File: instrumentation/httpclient5/src/test/java/brave/httpclient5/ITTracingHttpAsyncClientBuilder.java
Comment (adriancole) [inline instrumentation/httpclient5/src/test/java/brave/httpclient5/ITTracingHttpAsyncClientBuilder.java:53]: same about wide catch

File: pom.xml
Comment (adriancole) [inline pom.xml:258]: move these into httpclient5 pom as we ran into trouble sometimes controlling versions in parent. you can leave the property here or move it to httpclient5

**Review by adriancole — COMMENTED** [review]
Thanks for the big work koji! I made a first pass on some style things some maybe you might remember how we are weird..

File: instrumentation/httpclient5/src/main/java/brave/httpclient5/AsyncHandleSendHandler.java
Comment (kojilin) [reply instrumentation/httpclient5/src/main/java/brave/httpclient5/AsyncHandleSendHandler.java]: copy from httpclient(4.x) one, but looks like HttpClientHandler handles this well. Maybe I can add another PR to change the httpclient(4.x) later.

File: instrumentation/httpclient5/src/main/java/brave/httpclient5/TracingHttpClients.java
Comment (kojilin) [reply instrumentation/httpclient5/src/main/java/brave/httpclient5/TracingHttpClients.java]: Curious usually `HttpTracing` has the same server name, so the same instance should be used for the same endpoint? But separate to `newBuilder` & `build` still better than a single call.

File: instrumentation/httpclient5/src/test/java/brave/httpclient5/ITTracingCachingH2ClientBuilder.java
Comment (kojilin) [reply instrumentation/httpclient5/src/test/java/brave/httpclient5/ITTracingCachingH2ClientBuilder.java:30]: This is a test code, so I should keep it as public?

File: instrumentation/httpclient5/pom.xml
Comment (kojilin) [reply instrumentation/httpclient5/pom.xml]: it's 1.7, https://github.com/apache/httpcomponents-client/blob/master/README.txt thanks.

**Comment by adriancole** [comment]
do you mind to rebase this on latest master (note current version changed)

travis hasn't worked for almost 2 weeks. master now uses GitHub actions

**Comment by kojilin** [comment]
Updated, but 5.0.4 is not released yet. So I think we can merge after its release.

File: instrumentation/httpclient5/pom.xml
Comment (adriancole) [inline instrumentation/httpclient5/pom.xml:35]: Add XML comment here saying 5.0.4 is needed and ref the two issues. That comment should stay after we update so that people know (in case someone changes version and has a problem and no one remember)

File: instrumentation/httpclient5/src/main/java/brave/httpclient5/HttpClient5Tracing.java
Comment (adriancole) [inline instrumentation/httpclient5/src/main/java/brave/httpclient5/HttpClient5Tracing.java:73]: I would suggest name build for this method and above as code you are changing would just be calling builder.build() instead. also simplify by naming the param builder, (also in method above this)

File: instrumentation/httpclient5/src/main/java/brave/httpclient5/HttpClient5Tracing.java
Comment (adriancole) [inline instrumentation/httpclient5/src/main/java/brave/httpclient5/HttpClient5Tracing.java:87]: instead of this, make HttpClientUtils.closeScope/openScope a singleton. This also allows you to give the class a name which can be helpful in troubleshooting vs a lambda

I don't know the function addResponseInterceptorLast accepts, but say..

```java
// top-level file, but package private
final class TraceContextInterceptor implements RequestInterceptor, ResponseInterceptor {
  final CurrentTraceContext currentTraceContext;
 --snip--
```

Then, when you make new instance of HttpClient5Tracing, you save one of these as a field, and use it everywhere. it makes the instantiation look cleaner also.

File: instrumentation/httpclient5/src/main/java/brave/httpclient5/TracingHttpAsyncClient.java
Comment (adriancole) [inline instrumentation/httpclient5/src/main/java/brave/httpclient5/TracingHttpAsyncClient.java:75]: maybe swap this reference like this above call to `delegate.execute` as it is a little complicated. this allows you to ask why.. another option is to do `TraceContextFutureCallback.create` and make that function exit with input callback based on condition. also clean and comment can go there.

File: instrumentation/httpclient5/src/test/java/brave/httpclient5/ITTracingCachingH2ClientBuilder.java
Comment (adriancole) [reply instrumentation/httpclient5/src/test/java/brave/httpclient5/ITTracingCachingH2ClientBuilder.java:32]: sorry yeah we are still junit 4. my mistake

File: instrumentation/httpclient5/src/main/java/brave/httpclient5/HttpResponseWrapper.java
Comment (adriancole) [inline instrumentation/httpclient5/src/main/java/brave/httpclient5/HttpResponseWrapper.java:21]: you can find some wrapper tests in brave to copy/paste find/replace for these..  in async it is better to have unit tests for conversions as NPEs later in integration tests can be nasty

**Review by adriancole — COMMENTED** [review]
Looks very good Koji!

**Comment by adriancole** [comment]
fyi it seems both issues are marked fixed in 5.0.4, but I looked and there is no vote thread for that version. Hence, it might be until next year https://mail-archives.apache.org/mod_mbox/hc-dev/202012.mbox/browser

**Comment by skastenholz** [comment]
I am looking for a way to integrate Spring 5 Webclient and think it would be feasible with this "new" instrumentation. Any reason it did not yet get merged? Any alternative to instrument Spring 5 Webclient? Thank you.

**Comment by ParaSait** [comment]
Can this be merged, please? Is there anything still holding this back?

**Comment by jcchavezs** [comment]
Ping @shakuzen

**Comment by laurensdeb** [comment]
Has there been any progress towards merging this PR?

**Comment by kojilin** [comment]
Updated, I think all the tests are passed. Maybe need to check more due to I made it long before.

**Comment by laurensdeb** [comment]
@kojilin Any updates on this? We would love to integrate with HC 5 in one of our projects.

**Comment by kojilin** [comment]
I'm not the maintainer of the project, so need the other people to check and merge.

**Comment by jeqo** [comment]
I can add a review over the week. 

If some lone can earlier, please go ahead.
Sorry this has take this long

File: instrumentation/httpclient5/pom.xml
Comment (jeqo) [reply instrumentation/httpclient5/pom.xml]: @kojilin do you think the comments suggested here are worth adding?

File: instrumentation/httpclient5/src/main/java/brave/httpclient5/HttpClientUtils.java
Comment (jeqo) [inline instrumentation/httpclient5/src/main/java/brave/httpclient5/HttpClientUtils.java:56]: Hit the following issue when testing this instrumentation: https://github.com/openzipkin/brave-example/pull/91/files#r1159109251

I guess not all users have cache dependency in the classpath and will fail when this is evaluated on the response path. Could we handle this exception?

Comment (kojilin) [reply instrumentation/httpclient5/src/main/java/brave/httpclient5/HttpClientUtils.java:56]: Thanks, will handle this.

File: instrumentation/httpclient5/pom.xml
Comment (kojilin) [reply instrumentation/httpclient5/pom.xml]: I added the comment we need to use >= 5.2

**Review by jeqo — APPROVED** [review]
LGTM, thanks @kojilin !

**Comment by jeqo** [comment]
Seems license headers are up again, please run: `./mvnw com.mycila:license-maven-plugin:format` to fix it

**Comment by jcchavezs** [comment]
Any feedback here @shakuzen ?

**Review by shakuzen — APPROVED** [review]

**Review by jcchavezs — APPROVED** [review]

**Comment by jeqo** [comment]
Beautiful, thanks @kojilin !! 🎉

**Comment by JochemKuijpers** [comment]
Amazing work! 👏
This is the last blocker for an upgrade to Apache HTTP 5.x for me. 

@jeqo I noticed that after the merge, the CI tests failed: https://github.com/openzipkin/brave/actions/runs/4724094424
Seems something related to the license header that's not up to spec, but I'm not familiar with how this is supposed to run.

(how) can I help fix these issues so we can get a new artifact with the added HttpClient 5.2+ support?

**Pull request merged**
